agent/structs/config_entry_gateways.go

```diff
@@ -980,10 +980,130 @@ func (e *BoundAPIGatewayConfigEntry) GetEnterpriseMeta() *acl.EnterpriseMeta {
 	return &e.EnterpriseMeta
 }
 
+func (e *BoundAPIGatewayConfigEntry) UpdateRouteBinding(refs []ResourceReference, route BoundRoute) (bool, map[ResourceReference]error) {
+	didUpdate := false
+	errors := make(map[ResourceReference]error)
+
+	if len(e.Listeners) == 0 {
+		for _, ref := range refs {
+			errors[ref] = fmt.Errorf("route cannot bind because gateway has no listeners")
+		}
+		return false, errors
+	}
+
+	for i, listener := range e.Listeners {
+		// Unbind to handle any stale route references.
+		didUnbind := listener.UnbindRoute(route)
+		if didUnbind {
+			didUpdate = true
+		}
+		e.Listeners[i] = listener
+
+		for _, ref := range refs {
+			didBind, err := e.BindRoute(ref, route)
+			if err != nil {
+				errors[ref] = err
+			}
+			if didBind {
+				didUpdate = true
+			}
+		}
+	}
+
+	return didUpdate, errors
+}
+
+func (e *BoundAPIGatewayConfigEntry) BindRoute(ref ResourceReference, route BoundRoute) (bool, error) {
+	if ref.Kind != APIGateway || e.Name != ref.Name || !e.EnterpriseMeta.IsSame(&ref.EnterpriseMeta) {
+		return false, nil
+	}
+
+	if len(e.Listeners) == 0 {
+		return false, fmt.Errorf("route cannot bind because gateway has no listeners")
+	}
+
+	didBind := false
+	for i, listener := range e.Listeners {
+		if listener.Name == ref.SectionName || ref.SectionName == "" {
+			if listener.BindRoute(route) {
+				didBind = true
+				e.Listeners[i] = listener
+			}
+		}
```

Comment:
Somewhere in this conditional or in the `listener.BindRoute` implementation we should be checking that the protocol of the route aligns with the protocol type of the listener itself, i.e. `ListenerProtocolHTTP`-based listeners should only accept `HTTPRoute`s or error if binding, `ListenerProtocolTCP` should only accept `TCPRoute`.

Reply:
@t-eckert and I discussed this outside of this PR last week and this morning. It sounds like it would require some additional modifications to have the protocol readily accessible

> **Thomas Eckert**
> There will be some things to double back on that where we do the binding currently regardless of the protocol of the route matching the protocol of the listener, but I wanted to unblock everyone
> 
> And I want to talk to Andrew about adding the protocol to the BoundListener struct because without it we have to pause and go check the store every time we want to bind.

Reply:
Sounds good, I chatted a bit with Thomas already and think we're good with just throwing a TODO here and circling back on a subsequent lightweight refactor to add the protocol checking code.

My main thing is that I don't want to bloat the object we're actually persisting for state to avoid some complicated deprecation/refactoring/migrations later, so I'd prefer not to store the protocol on the `BoundListener`. Instead, I'm thinking we might want to do some sort of tuple wrapper similar to what we have in our k8s controller code, where we have an intermediate `K8sGateway` that wraps both our computed state and the actual Gateway configuration and abstracts away a bit of the interaction between the two. But, that can be left for the follow-up.

Reply:
Added the TODO